File: server-kotlin/src/main/kotlin/com/example/emojichat/RoomController.kt
package com.example.emojichat

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.UUID

@RestController
@RequestMapping("/api/rooms")
class RoomController(private val repo: ChatRepository) {

    @PostMapping
    fun create(@RequestBody req: CreateRoomRequest): ResponseEntity<Any> {
        val ownerId = req.ownerId?.takeIf { it.isNotBlank() }
            ?: return error(HttpStatus.BAD_REQUEST, "owner_id required")
        val room = repo.createRoom(
            id = UUID.randomUUID().toString(),
            ownerId = ownerId,
            name = req.name ?: "",
        )
        return ResponseEntity.status(HttpStatus.CREATED).body(room)
    }

    @GetMapping("/{id}")
    fun get(@PathVariable id: String): ResponseEntity<Any> {
        val room = repo.findRoom(id) ?: return error(HttpStatus.NOT_FOUND, "room not found")
        return ResponseEntity.ok(room)
    }

    @DeleteMapping("/{id}")
    fun delete(
        @PathVariable id: String,
        @RequestHeader(name = "X-User-Id", required = false) userId: String?,
    ): ResponseEntity<Any> {
        if (userId.isNullOrBlank()) return error(HttpStatus.BAD_REQUEST, "X-User-Id header required")
        val owner = repo.findOwner(id) ?: return error(HttpStatus.NOT_FOUND, "room not found")
        if (owner != userId) return error(HttpStatus.FORBIDDEN, "only the room owner can delete")
        repo.deleteRoom(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/{id}/messages")
    fun messages(
        @PathVariable id: String,
        @RequestParam(required = false) limit: Int?,
        @RequestParam(required = false) before: Long?,
    ): Map<String, Any> {
        val safeLimit = (limit ?: 50).coerceIn(1, 200)
        return mapOf("messages" to repo.listMessages(id, safeLimit, before))
    }

    private fun error(status: HttpStatus, msg: String): ResponseEntity<Any> =
        ResponseEntity.status(status).body(mapOf("error" to msg))
}

@RestController
class RootController {
    @GetMapping("/")
    fun root(): ResponseEntity<Void> =
        ResponseEntity.status(HttpStatus.FOUND)
            .header("Location", "/static/index.html")
            .build()
}

File: server-elixir/lib/emoji_chat_web/controllers/room_controller.ex
defmodule EmojiChatWeb.RoomController do
  use EmojiChatWeb, :controller

  alias EmojiChat.Chat

  def create(conn, params) do
    owner_id = trim(params["owner_id"])

    if owner_id == "" do
      json_error(conn, 400, "owner_id required")
    else
      room = Chat.create_room(owner_id, params["name"])
      conn |> put_status(201) |> json(room)
    end
  end

  def show(conn, %{"id" => id}) do
    case Chat.get_room(id) do
      nil -> json_error(conn, 404, "room not found")
      room -> json(conn, room)
    end
  end

  def delete(conn, %{"id" => id}) do
    user_id = trim(get_req_header(conn, "x-user-id") |> List.first())

    cond do
      user_id == "" ->
        json_error(conn, 400, "X-User-Id header required")

      true ->
        case Chat.get_owner(id) do
          nil ->
            json_error(conn, 404, "room not found")

          owner when owner != user_id ->
            json_error(conn, 403, "only the room owner can delete")

          _ ->
            Chat.delete_room(id)
            conn |> send_resp(204, "")
        end
    end
  end

  defp trim(nil), do: ""
  defp trim(s) when is_binary(s), do: String.trim(s)

  defp json_error(conn, status, msg) do
    conn |> put_status(status) |> json(%{error: msg})
  end
end
